### ISSUE TITLE
(M) As a developer I would like a unit test suite for the current securitas codebase

### ISSUE THREAD

**Issue by sfinn85** [fedora-infra/noggin#54]
Proof of concept tests 

**Acceptance Criteria:** 
Creation of integration & unit tests 
Coverage 

**DOD:** 
Tests to be completed identified and documented
Reviewed by team/sign off by team

**Comment by puiterwijk** [comment]
As an AAA developer I want clear descriptions of my issues.

**Comment by ryanlerch** [comment]
any thoughts on how to approach unit tests for the web front end here? since we kind of require an IPA server for it to work properly.

**Comment by abompard** [comment]
I see two options:
- We mock the client that python-freeipa gives us, and in each test we use `return_value` and `side_effect` to make it return what we want to test. If we choose that we'll probably end up building a mocking client class
- We use [python-vcr](https://vcrpy.readthedocs.io) to record and replay the HTTP requests. That means that we need to have a working freeipa server when the unit test is first written, but I think that's a good thing because with mocking there's always the risk that the return value you set is not something that can actually happen and then you end up testing against a false service and your test is useless.

Mocking the client is easier to start with but I think VCR can be easier in the long run.

**Comment by ryanlerch** [comment]
Thanks @abompard!

python-vcr seems like it might be the way to go here. However, still kind of blocked on having a running free-ipa instance i can point to to start writing these tests.

**Comment by StephenCoady** [comment]
@ryanlerch I managed to get freeipa running in a container. I'm still a little bit off having a full local dev environment but if all you need is a freeipa server running locally then the following should get you there. The steps are:

clone this repo: https://github.com/freeipa/freeipa-container
from within run `podman build -t freeipa-server .`
run the following: `setsebool -P container_manage_cgroup 1`
run `sudo mkdir /var/lib/ipa-data`
to bring up the container run 
```
podman run --name freeipa -ti -p 80:80 -p 443:443 \
   -h ipa.example.test \
   -v /sys/fs/cgroup:/sys/fs/cgroup:ro \
   --tmpfs /run --tmpfs /tmp \
   -v /var/lib/ipa-data:/data:Z freeipa-server
```
configure the above with the defaults it suggests except for the ca cert suggestion where you should type yes
add the following to your /etc/hosts file `127.0.0.1 ipa.example.test`
once done you should be able to go to localhost and be sent to ipa.example.test

sorry its a mess i typed it in a rush before finishing I hope it works for you. I'm hoping to have it in a compose file tomorrow so that should make things easier.

**Comment by sfinn85** [comment]
Great work all, keep me updated on this. If there is anything I can possibly help with.. let me know. :)

**Comment by abompard** [comment]
Update on this, PR #81 brings two controllers to 100% coverage (`authentication` and `registration`), but there are more to go, and other components that are not controllers and that are just as important (especially the `security` module).

I'm pretty sure that this story is going to take much longer than we estimated and I think it should be noted. Much more than L-size. I've spent 2 days covering two controllers, and even though it should be faster now because I did some foundation work, it's still going to take some time to cover the nominal and the error paths.

On the other hand, it's a good thing: it's because the current securitas codebase does a lot of things that we have a lot of testing to do to cover it ;-)
